src/reaction.cpp

```diff
@@ -174,32 +223,98 @@ const std::unordered_map<int, std::string> REACTION_NAME_MAP {
   {HEATING_LOCAL, "heating-local"},
 };
 
+std::unordered_map<std::string, int> REACTION_TYPE_MAP;
+
+void initialize_maps()
+{
+  // Add level reactions to name map
+  for (int level = 0; level <= 48; ++level) {
+    if (level >= 1 && level <= 40) {
+      REACTION_NAME_MAP[50 + level] = fmt::format("(n,n{})", level);
+    }
+    REACTION_NAME_MAP[600 + level] = fmt::format("(n,p{})", level);
+    REACTION_NAME_MAP[650 + level] = fmt::format("(n,d{})", level);
+    REACTION_NAME_MAP[700 + level] = fmt::format("(n,t{})", level);
+    REACTION_NAME_MAP[750 + level] = fmt::format("(n,3He{})", level);
+    REACTION_NAME_MAP[800 + level] = fmt::format("(n,a{})", level);
+    if (level <= 15) {
+      REACTION_NAME_MAP[875 + level] = fmt::format("(n,a{})", level);
+    }
+  }
+
+  // Create photoelectric subshells
+  for (int mt = 534; mt <= 572; ++mt) {
+    REACTION_NAME_MAP[mt] = fmt::format("photoelectric, {} subshell",
+      SUBSHELLS[mt - 534]);
+  }
+
+  // Invert name map to create type map
+  for (const auto& kv : REACTION_NAME_MAP) {
+    REACTION_TYPE_MAP[kv.second] = kv.first;
+  }
+}
+
 std::string reaction_name(int mt)
 {
-  if (N_N1 <= mt && mt <= N_N40) {
-    return fmt::format("(n,n{})", mt - 50);
-  } else if (534 <= mt && mt <= 572) {
-    return fmt::format("photoelectric, {} subshell", SUBSHELLS[mt - 534]);
-  } else if (N_P0 <= mt && mt < N_PC) {
-    return fmt::format("(n,p{})", mt - N_P0);
-  } else if (N_D0 <= mt && mt < N_DC) {
-    return fmt::format("(n,d{})", mt - N_D0);
-  } else if (N_T0 <= mt && mt < N_TC) {
-    return fmt::format("(n,t{})", mt - N_T0);
-  } else if (N_3HE0 <= mt && mt < N_3HEC) {
-    return fmt::format("(n,3He{})", mt - N_3HE0);
-  } else if (N_A0 <= mt && mt < N_AC) {
-    return fmt::format("(n,a{})", mt - N_A0);
-  } else if (N_2N0 <= mt && mt < N_2NC) {
-    return fmt::format("(n,2n{})", mt - N_2N0);
+  // Initialize remainder of name map and all of type map
+  if (REACTION_TYPE_MAP.empty()) initialize_maps();
+
+  // Get reaction name from map
+  auto it = REACTION_NAME_MAP.find(mt);
+  if (it != REACTION_NAME_MAP.end()) {
+    return it->second;
   } else {
-    auto it = REACTION_NAME_MAP.find(mt);
-    if (it != REACTION_NAME_MAP.end()) {
-      return it->second;
-    } else {
-      return fmt::format("MT={}", mt);
-    }
+    return fmt::format("MT={}", mt);
+  }
+}
+
+int reaction_type(std::string name)
+{
+  // Initialize remainder of name map and all of type map
+  if (REACTION_TYPE_MAP.empty()) initialize_maps();
+
+  // Check if type map has an entry for this reaction name
+  auto it = REACTION_TYPE_MAP.find(name);
+  if (it != REACTION_TYPE_MAP.end()) {
+    return it->second;
+  }
+
+  // Alternate names for several reactions
+  if (name == "(n,total)") {
```

Comment:
If these names are in reaction type map wont this routine have returned already?

Reply:
It looks like (n,total) needs to be handled separately actually. All the other names are alternate names that don't appear in the map, but (n,total) does appear, so I'm going to change this to check for it before the map.